beancount/parser/booking_full.py

```diff
@@ -195,24 +195,48 @@ def _book(entries, options_map, methods):
 
                 if interpolation_errors:
                     errors.extend(interpolation_errors)
+
+                # Update the running balances for each account using the
+                # booked and interpolated values. Note that we could optimize away
+                # some of this in book_reductions() but we choose not to do so, as a
+                # sanity check that the direct aggregation of the final booked lots
+                # will compute the same result as that during the book_reductions()
+                # process.
+                for posting in inter_postings:
+                    balance = balances[posting.account]
+                    balance.add_position(posting)
+
+                # If lot merging was requested, resolve it after interpolation.
+                # This allows interpolation to resolve an unspecified cost before
+                # it gets merged together into a merged inventory.
+                cost_merging_postings = []
+                for i, inter_posting in enumerate(inter_postings):
+                    # Interpolated postings are always Cost instead of CostSpec,
+                    # so we need to look back at the pre-interpolated postings
+                    # to check the merge flag.
+                    posting_cost = booked_postings[i].cost
+                    if posting_cost is None:
+                        continue
+                    if (isinstance(posting_cost, CostSpec) and posting_cost.merge) \
+                            or methods[inter_posting.account] is Booking.AVERAGE:
+                        cost_merging_postings.extend(
+                            booking_method.rebook_inventory_at_average_cost(
+                                balances[inter_posting.account],
+                                inter_posting.account,
+                                inter_posting.units.currency,
+                                inter_posting.cost.currency,
+                            )
+                        )
+
                 repl_postings.extend(inter_postings)
+                repl_postings.extend(cost_merging_postings)
```

Comment:
The first line could move to before the block; the second one would be within the special loop and conditionally extend.
You could avoid 'cost_merging_postings' entirely.